[BEAM-357] fixing build on windows

rmannibucau opened the pull request:
Small fixes or workarounds for windows.

Main issues were:
- Paths.get(...): windows has ":" which will make it failing, using new File(...).toPath() to fix it
- Workaround for hadoop which requires winutils.exe on windows (see https://issues.apache.org/jira/browse/BEAM-357 for a comment on it)

jbonofre commented:
R: @jbonofre

jbonofre commented:
Starting the review.

davorbonaci commented on sdks/java/maven-archetypes/starter/pom.xml:
```diff
@@ -60,6 +60,9 @@
                 <goal>integration-test</goal>
               </goals>
+              <configuration>
+                <ignoreEOLStyle>true</ignoreEOLStyle> <!-- for win -->
```
Can you remove the comment? Alternatively, please expand as to why it is needed, etc.

davorbonaci commented on sdks/java/io/hdfs/pom.xml:
```diff
@@ -83,5 +83,14 @@
       <scope>test</scope>
     </dependency>
+
+    <!-- see HDFSFileSourceTest commented block
```
I'm not sure what comment block are you referring to. Either way, I'd remove the commented dependency.

davorbonaci commented on sdks/java/core/src/main/java/org/apache/beam/sdk/testing/HadoopWorkarounds.java:
```diff
@@ -0,0 +1,129 @@
+/**
+ * A simple class ensure winutils.exe can be found in the JVM.
+ */
+public class HadoopWorkarounds {
```
We cannot really download random, untrusted files and execute them on users' machine. We have to find a different way of solving this problem.

davorbonaci commented:
Hi Romain,
Welcome to Beam! I'm glad to see your contribution!

On a high level -- we really need to work well on Windows. Thanks for starting this effort. In fact, we need to add integration testing on Windows as soon as possible.

Code-wise, we need to find a different solution -- we cannot really download random, untrusted files and execute them on users' machine.

Again, thanks and welcome!

rmannibucau commented:
Hi Davor,

Updated the comments. Let me know if it is better or worse ;).

Changed a bit the downloading to add a sanity check (we could replace it by a checksum but the best would be to be able to use gpg --verify but seems bouncycastle doesn't fully match gpg format).

This repo is not really random since it is provided by an asf committer and I downgraded to a signed binary version. A maybe better alternative is to package a beam-hadoop-win jar with this exe and just explode it if running tests on windows - drawback is we put a binary in a jar, advantage is beam owns this binary.

wdyt?

davorbonaci commented:
This is definitely safer, but I feel we are still exposing a potentially unnecessary attack surface to our users.

Would you be able to summarize the problem on the Windows platform, and how does this binary solves it?

rmannibucau commented:
Sure, on windows hadoop relies for most commands (including chmods which are the first commands triggered in beam) on winutils.exe to actually execute the action (https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/Shell.java#L224 for instance). It does find the winutils.exe from hadoop home which is ATM not set in beam. If you have locally hadoop with winutils.exe - not provided by ASF yet - and set the home properly then most of tests - excepted the one using Path but that's something else - should be fine but if you don't you need that workaround cause Shell of hadoop get winutils in a static block (https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/Shell.java#L639) and it can be null. If null it will not remove the option but execute commands with a null entry in the array (processbuilder.execute([null, "chmod", ....])) which leads to a NPE.

Side note: if you are curious chmod is implemented like that https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-common/src/main/winutils/chmod.c with win API

We can also propose to hadoop a degraded mode where if winutils is not there and a "test" flag activated we try to do these operations in java and if it fails we ignore them when acceptable (chmod/chown should be ok to ignore for tests for instance). Didn't check if it would be doable for all winutils operations.

davorbonaci commented:
I think the alternative could be a better approach overall. In general, `chmod` is rarely needed on Windows and with minimal benefits in the testing environment.

There are others too -- tools like Cygwin, etc. provide many of these utilities. Finally, Microsoft announced [Windows Subsystem for Linux (WSL)](https://blogs.windows.com/buildingapps/2016/03/30/run-bash-on-ubuntu-on-windows/) that enables the scenario too.

So, I think we could try with a degraded mode first. What do you think?

Separately, I have to kindly ask not to [push commits](https://lists.apache.org/thread.html/b74fd53d19088cf37561f411c6bd2f41e6c3e9db076d69dfb280640c@%3Ccommits.beam.apache.org%3E) to our repository, including creating a [feature branch](https://lists.apache.org/thread.html/853a948e571c36c1e819ab704918340a091fcc664f73c4fab61b0063@%3Ccommits.beam.apache.org%3E).

rmannibucau commented:
Yep, cygwin could be detected and surely enough.

Will try to find some time for it and finally change this PR for a hadoop upgrade + the few other fixes.

About the push: mea culpa, bad ctrl+R, deleted it just after.

davorbonaci commented:
(Also, feel free to separate into multiple PR. A good portion of this one can probably go in separately and is non-controversial. Of course, do as you see fit.)

jbonofre commented:
Rebasing and checking ...

davorbonaci commented:
Should we close this if it is not active, and track it in JIRA instead, which can link back to the closed PR?

rmannibucau commented:
ok with me

jbonofre commented:
+1 to close

davorbonaci commented:
@rmannibucau, can you close please then?

rmannibucau commented:
yep :)
